library/pod-security-policy/users/mutation-mustRunAsNonRoot.yaml

```diff
@@ -0,0 +1,18 @@
+apiVersion: mutations.gatekeeper.sh/v1alpha1
+kind: Assign
+metadata:
+  name: k8spsprunasnonroot
+spec:
+  applyTo:
+  - groups: [""]
+    kinds: ["Pod"]
+    versions: ["v1"]
+  location: "spec.containers[name:*].securityContext.runAsNonRoot"
+  parameters:
+    pathTests:
+    # - subPath: "spec.containers[name:*].securityContext.runAsUser"
```

Comment:
Is there any harm in setting runAsNonRoot even if runAsUser is also set?

Reply:
Kubelet gives `CreateContainerConfigError` if they are conflicting: (runAsUser=0 and runAsNonRoot=true)

```
  Warning  Failed     17s                kubelet            Error: container's runAsUser breaks non-root policy (pod: "nginx-run-as-root_default(350bd4bc-10c2-4299-912f-a9651dfbadab)", container: nginx)
```

Reply:
I'm thinking this is a job for the match criteria.

If you have a constraint that is enforcing run-as-non-root, you'll want this defaulting set, at which point you'd like the request to be rejected anyway (though it would be nice to know which would win, the violation message or the conflict message, as the former is more user-friendly).

If users want to allow run as root, then they probably have no opinion as to whether/how `runAsNonRoot` is set.

If they want to force run as root to be set, they would have two mutations so they could default both values.

This does point to a general syntactic weakness when dealing with [union](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#unions) type objects, however.

Reply:
Another option would be to default the `securityContext` object in its entirety

Reply:
Wouldn't setting `securityContext` be same issue because of non-prefix path tests?

What is the next step here? Should we remove the commented out part and leave it up to validation to check for `runAsUser` to be non-zero?

Reply:
The two different paths share a common root at `spec.containers[name:*].securityContext`

Yeah, we can delete the comment, or rejigger the example to avoid the potential issue.

Reply:
```
  location: "spec.containers[name:*].securityContext"
  parameters:
    pathTests:
    - subPath: "spec.containers[name:*].securityContext.runAsUser"
      condition: MustNotExist
    - subPath: "spec.containers[name:*].securityContext.runAsNonRoot"
      condition: MustNotExist
    assign:
      value:
        runAsNonRoot: true
```

like this? this also gives non-prefix error

I removed the path test for now.

Reply:
More like this:

```
  location: "spec.containers[name:*].securityContext"
  parameters:
    assign:
      value:
        runAsNonRoot: true
```

The path test is unnecessary as the glob expansion won't take effect if there are no containers, and we want to set all security contexts to "run as non-root"

Reply:
> we want to set all security contexts to "run as non-root"

If we want similar functionality with PSP, it's set all security contexts unless it's defined explicitly.

Example scenario:
- User has a certain workload (example, uses port 80) that can run as root only, so user sets it accordingly
- Gatekeeper won't mutate since path tests are set (otherwise, workload won't run at all)
- User adds exclusion for that workload so it's not denied, but rest of the workloads are still part of mutation and validation

Reply:
Then you could add a pathTest like this:

```
  location: "spec.containers[name:*].securityContext"
  parameters:
    pathTests:
    - subPath: "spec.containers[name:*].securityContext"
      condition: MustNotExist
    assign:
      value:
        runAsNonRoot: true
```

Reply:
```
    - subPath: "spec.containers[name:*].securityContext"
      condition: MustNotExist
```

I think this path test is too broad since securityContext can have other things like `capabilities` that's not related to users.

Reply:
IMO that's up to the admin... I don't think this can be one size fits all.

Reply:
Also, because it's saying `MustNotExist`, it won't override any security context provided by the user